Handle existing scheme in host

thomashacker opened the pull request:
This `PR` aims to fix https://github.com/weaviate/typescript-client/issues/87 and adds a simple check before creating the `baseUrl`

```
const version = '/v1';
const baseUri = config.host.startsWith(`${config.scheme}://`)
    ? `${config.host}${version}`
    : `${config.scheme}://${config.host}${version}`;
```

This allows users to use a host URL with a scheme if it matches the specified scheme.

The PR also adds two unit tests to verify that it constructs the correct URLs.

tsmith023 reviewed (commented):
Thanks for the PR @thomashacker 😁

Shall we make the `config.host` parameter optional now that users are free to supply it within the host? Then the scheme would be pulled out of `config.host` by splitting the string on `://`. If they don't provide `config.scheme` and it cannot be found in `config.host`, then an error is thrown.

I'm wondering if it would ever be confusing for users coming from Python to specify the URL in `config.host` including the protocol only to then have to specify it in `config.scheme` also. What do you think?

thomashacker commented:
I agree,

As far as I can see, the `scheme` parameter is only being used to build the `baseURL` for both the `HTTPClient` and `GQLClient`. As you suggested, If the scheme is already present in the host, we can make specifying the scheme optional.

I would also move the whole `scheme`/`host` processing upwards, I'll look into it!

thomashacker commented:
- Made the `scheme` parameter optional and removed `throw error` when it doesn't exist
- Moved `schema` and `host` checks to `Connection`, adding following unit tests:
1. Host contains scheme and it's the same as specified (Use host only)
2. Host contains scheme but it's not the same as specified (throw error)
3. Host contains scheme and scheme was not specified (Use host only)
4. Host does not contain scheme and no scheme was specified (throw error)
5. Host does not contain scheme but a scheme was specified (default)

- Both the `HttpClient` and `GQLClient` are now using `params.host` parameter + adding their prefixes

tsmith023 commented on src/connection/unit.test.ts:
```diff
@@ -124,6 +124,75 @@ describe('mock server auth tests', () => {
+    };
+
+    expect(createConnection).toThrow(
+      'The host contains a different protocol than specified in the scheme (scheme: https:// != host: http://)'
```
On a brief glance, I got a bit confused by this error message. Might it be more accurate to say:
`'The host contains a different protocol than specified in the scheme (scheme: https != host: http)'`?

I can foresee a user seeing this error and trying to add `https://` to the `scheme` field to fix it

thomashacker replied:
Good idea, I changed the error message to make it more clear!

tsmith023 reviewed (approved):
LGTM otherwise :)
